### ISSUE TITLE
cycle detected loading repository directory '@com_zombiezen_go_capnproto2'

### ISSUE THREAD

**Issue by junghoahnsc** [bazelbuild/rules_go#1551]
I'm trying to use `go_repository` from `@bazel_gazelle//:def.bzl`, since it's deprecated.
But I got the following error:

```
$ bazel build  @com_zombiezen_go_capnproto2//...
ERROR: cycle detected loading repository directory '@com_zombiezen_go_capnproto2'
ERROR: cycles detected during target parsing
```

WORKSPACE looks like this:
```
git_repository(
    name = "io_bazel_rules_go",
    remote = "https://github.com/bazelbuild/rules_go.git",
    tag = "0.11.2",
)

git_repository(
    name = "bazel_gazelle",
    remote = "https://github.com/bazelbuild/bazel-gazelle.git",
    tag = "0.12.0",
)

go_repository(
        name = "com_zombiezen_go_capnproto2",
        importpath = "zombiezen.com/go/capnproto2",
        vcs = "git",
        remote = "https://github.com/capnproto/go-capnproto2",
        # The existing build file has no importpath causing build failure.
        # We force bazel to generate bazel BUILD file.
        build_file_name = "BUILD.bazel",
        build_file_generation = "on",
        tag = "v2.17.0",
    )
```

Is this a known issue?

**Comment by jayconrod** [comment]
Hmm, could you post your full WORKSPACE file? The one below works for me with the same build command.

I've opened PR capnproto/go-capnproto2#118 to fix this upstream. Also issue bazelbuild/bazel-gazelle#232 to let you replace existing build files in `go_repository`.

```
git_repository(
    name = "io_bazel_rules_go",
    remote = "https://github.com/bazelbuild/rules_go.git",
    tag = "0.11.2",
)

git_repository(
    name = "bazel_gazelle",
    remote = "https://github.com/bazelbuild/bazel-gazelle.git",
    tag = "0.12.0",
)

load("@io_bazel_rules_go//go:def.bzl", "go_register_toolchains", "go_rules_dependencies")

go_rules_dependencies()

go_register_toolchains()

load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies")

gazelle_dependencies()

load("@bazel_gazelle//:def.bzl", "go_repository")

go_repository(
    name = "com_zombiezen_go_capnproto2",
    build_file_generation = "on",
    # The existing build file has no importpath causing build failure.
    # We force bazel to generate bazel BUILD file.
    build_file_name = "BUILD.bazel",
    importpath = "zombiezen.com/go/capnproto2",
    remote = "https://github.com/capnproto/go-capnproto2",
    tag = "v2.17.0",
    vcs = "git",
)

go_repository(
    name = "com_github_kylelemons_godebug",
    commit = "d65d576e9348f5982d7f6d83682b694e731a45c6",
    importpath = "github.com/kylelemons/godebug",
)
```

**Comment by junghoahnsc** [comment]
Thanks for looking into this.

My `WORKSAPCE` is huge, but basically it looks like this:

```
git_repository(
    name = "io_bazel_rules_go",
    remote = "https://github.com/bazelbuild/rules_go.git",
    tag = "0.11.2",
)

git_repository(
    name = "bazel_gazelle",
    remote = "https://github.com/bazelbuild/bazel-gazelle.git",
    tag = "0.12.0",
)

load("@bazel_gazelle//:def.bzl", "go_repository")

# A bunch of go repositories.
go_repository(....)
...

load("@io_bazel_rules_go//go:def.bzl", "go_register_toolchains", "go_rules_dependencies")

go_rules_dependencies()

go_register_toolchains()

load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies")

gazelle_dependencies()
```

I intentionally load go-rule dependencies after loading my own repositories to control versions of thirdparty libs. I tried to find which lib causes this and found that the following repository was the issue.

```
    # TOML library.
    go_repository(
        name = "com_github_pelletier_go_toml",
        importpath = "github.com/pelletier/go-toml",
        tag = "v1.1.0",
    )
```

When I commented this out, it works. Do I need to handle this repository specially?

**Comment by jayconrod** [comment]
Interesting. `@com_github_pelletier_go_toml` was a dependency of Gazelle in the past. I vendored it a few months ago, but rules_go still declares it in `go_rules_dependencies` to make old versions of `go_repository` and the `gazelle` rule work. It will probably be removed soon.

I'm not sure where exactly the dependency cycle is coming from. I think some old version of Gazelle may be declared somewhere, and the rule used to declare `com_github_pelletier_go_toml` depends on that version of Gazelle via `go_repository`. I'd suggest reordering your WORKSPACE slightly to avoid overrides:

1.
```
git_repository(
    name = "io_bazel_rules_go",
    remote = "https://github.com/bazelbuild/rules_go.git",
    tag = "0.11.2",
)

git_repository(
    name = "bazel_gazelle",
    remote = "https://github.com/bazelbuild/bazel-gazelle.git",
    tag = "0.12.0",
)
```
Everything depends on these. It's important that `bazel_gazelle` is declared before `go_rules_dependencies`, since it may be declared inside that macro if it's not already defined.
2. 
```
load("@io_bazel_rules_go//go:def.bzl", "go_register_toolchains", "go_rules_dependencies")`
go_rules_dependencies()
go_register_toolchains()
```
3.
```
load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies", "go_repository")
gazelle_dependencies()
```
Note that `go_repository` can be loaded from `@bazel_gazelle//:deps.bzl`. I think it's safe to load it from either file, but if you load it from `def.bzl`, make sure `gazelle_dependencies()` has been called earlier.
4. All other `go_repository` rules.

**Comment by junghoahnsc** [comment]
Yes, it works if I called `go_rules_dependencies()` and `gazelle_dependencies()` before all other own `go_repository` rules. 

But in that case, we can't set the versions of other repositories that are already loaded by go/gazelle rule dependencies. Is there a way to set the versions in `go_rules_dependencies()` and `gazelle_dependencies()`?

**Comment by jayconrod** [comment]
You can still set the versions. Bazel repository rules are a little quirky this way. When you declare a repository rule with the same name as a previously declared repository rule it *may* override the previous rule, depending on the implementation of that rule.

If the previous rule was a native repository rule (`http_archive` or `git_repository`), the new rule will be silently ignored.

If the previous rule was a Skylark repository rule (`go_repository`), the new rule will override the old rule.

All rules declared inside `go_rules_dependencies` and `gazelle_dependencies` are Skylark rules, so you should be able to override them.

Note also that Bazel is deprecating the native `http_archive` and `git_repository` rules. There are Skylark replacements in `@bazel_tools//tools/build_defs/repo:http.bzl` and `@bazel_tools//tools/build_defs/repo:git.bzl` (`@bazel_tools` is a builtin, pre-declared repo).

**Comment by pmuetschard** [comment]
> You can still set the versions.

This does not appear to be true. What am I doing wrong? Given the following `WORKSPACE`:
```
git_repository(
    name = "io_bazel_rules_go",
    remote = "https://github.com/bazelbuild/rules_go.git",
    tag = "0.11.2",
)

git_repository(
    name = "bazel_gazelle",
    remote = "https://github.com/bazelbuild/bazel-gazelle.git",
    tag = "0.12.0",
)

load("@io_bazel_rules_go//go:def.bzl", "go_register_toolchains", "go_rules_dependencies")
go_rules_dependencies()
go_register_toolchains()

load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies", "go_repository")
gazelle_dependencies()

go_repository(
    name = "org_golang_x_tools",
    commit = "25b95b48224cce18163c7d49dcfb89a2d5ecd209",
    importpath = "golang.org/x/tools",
)
```
It still loads commit `5d2fd3ccab986d52112bf301d47a819783339d0e`, which is from `rules_go`:
```
INFO: SHA256 (https://github.com/bazelbuild/rules_go/archive/0.11.2.tar.gz) = 5a89e9c1fe89d45c4a45132fae94bf130842ce3c77e3a735a68f6991d4e00a25
INFO: SHA256 (https://codeload.github.com/golang/tools/zip/5d2fd3ccab986d52112bf301d47a819783339d0e) = 31c4b514df1b40d7ade2e6f824dca35a36d9454983e7955074ac6f47f1eb3292
```

**Comment by pmuetschard** [comment]
FYI, this seems to be `org_golang_x_tools` related, as I was successfully able to override `com_github_golang_glog`.

**Comment by jayconrod** [comment]
I ran a quick experiment. Previously, my understanding was that if when a repository is declared multiple times (assuming Skylark repository rule implementation), the last declaration wins. This is not entirely true. If something causes the repository to be fetched during WORKSPACE evaluation, the current repository declaration at the point where the fetch occurs will win. Later declarations will be silently ignored.

I *think* this has been fixed with the tip commits of rules_go and Gazelle. 

`@org_golang_x_tools` used to be a dependency of Gazelle. We used a package which has since been vendored. Gazelle is built inside a special repository used by `go_repository` so `@org_golang_x_tools` was probably fetched in order to build that.

The change is not part of an official release yet. It may be possible to do a point release with that change in Gazelle. I'll try that out early next week.
